File: src/speed2.cpp
// ==========================================
// Basic Libraries & Headers
// ==========================================
#include "rclcpp/rclcpp.hpp"
#include "rclcpp/node.hpp"
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <std_msgs/msg/float32.hpp>
#include <cmath>
#include <chrono>
#include <string>
#include <optional>
#include <unordered_map>
#include <stdexcept>
#include "heading_controller_pkg/speed2_utils.hpp"

// ==========================================
// Type Aliases for Standard ROS 2 Messages
// ==========================================
using NavSatFix = sensor_msgs::msg::NavSatFix;
using TwistStamped = geometry_msgs::msg::TwistStamped;
using Float32 = std_msgs::msg::Float32;

// ==========================================
// GPS Distance & Velocity Node Class
// (SPEED_CHART / SPEED_CHART_INVERSE / convert_speed() / get_distance()
// now live in gps_utils.hpp so they can be unit-tested directly with
// gtest, without needing a running rclcpp::Node)
// ==========================================
class GPSDistanceVelocityNode : public rclcpp::Node {
public:
    GPSDistanceVelocityNode() : Node("gps_distance_velocity_node") {
        // ----------------------------------------------------
        // Initialize State Variables
        // ----------------------------------------------------
        prev_lat_ = std::nullopt;
        prev_lon_ = std::nullopt;
        total_distance_ = 0.0f;
        gps_abs_vel_ = 0.0f;
        car_speed_kmh_ = 0.0f;

        // ----------------------------------------------------
        // Setup Subscriptions and Publications
        // ----------------------------------------------------
        speed_pub_ = this->create_publisher<Float32>("/speed", 10);
        speed_pub_mps_ = this->create_publisher<Float32>("/speed_mps", 10);

        gps_sub_ = this->create_subscription<NavSatFix>(
            "/fix", 10,
            std::bind(&GPSDistanceVelocityNode::gps_callback, this, std::placeholders::_1)
        );

        gps_vel_sub_ = this->create_subscription<TwistStamped>(
            "/vel", 10,
            std::bind(&GPSDistanceVelocityNode::gps_vel_callback, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "✅ GPS Distance & Velocity Node Started");

        // ----------------------------------------------------
        // Setup Periodic Timer for Summary Logs
        // ----------------------------------------------------
        timer_ = this->create_wall_timer(
            std::chrono::duration<float>(2.0f),
            std::bind(&GPSDistanceVelocityNode::timer_callback, this)
        );
    }

private:
    // ==========================================
    // Callback: GPS Position
    // ==========================================
    void gps_callback(const NavSatFix::SharedPtr msg) {
        double lat = msg->latitude;
        double lon = msg->longitude;

        if (prev_lat_.has_value() && prev_lon_.has_value()) {
            double dist = get_distance(prev_lat_.value(), prev_lon_.value(), lat, lon);
            total_distance_ += static_cast<float>(dist);
            RCLCPP_INFO(this->get_logger(),
                        "📍 Distance from last point: %.3f m  |  Total: %.3f m",
                        dist, total_distance_);
        }

        prev_lat_ = lat;
        prev_lon_ = lon;
    }

    // ==========================================
    // Callback: GPS Velocity
    // ==========================================
    void gps_vel_callback(const TwistStamped::SharedPtr msg) {
        float vel_x = static_cast<float>(msg->twist.linear.x);
        float vel_y = static_cast<float>(msg->twist.linear.y);

        // Ground speed (horizontal magnitude)
        gps_abs_vel_ = std::sqrt(vel_x * vel_x + vel_y * vel_y);

        // Convert m/s → km/h
        try {
            car_speed_kmh_ = convert_speed(gps_abs_vel_, "m/s", "km/h");
        } catch (...) {
            car_speed_kmh_ = 0.0f;
        }

        // Publish speed in m/s and km/h
        Float32 speed_msg_mps;
        speed_msg_mps.data = gps_abs_vel_;
        speed_pub_mps_->publish(speed_msg_mps);

        Float32 speed_msg_kmh;
        speed_msg_kmh.data = car_speed_kmh_;
        speed_pub_->publish(speed_msg_kmh);

        RCLCPP_INFO(this->get_logger(),
                    "🚗 Ground Speed: %.3f m/s  |  Car Speed: %.3f km/h",
                    gps_abs_vel_, car_speed_kmh_);
    }

    // ==========================================
    // Timer: Periodic summary
    // ==========================================
    void timer_callback() {
        RCLCPP_INFO(this->get_logger(),
                    "⏱️ Speed: %.2f m/s  (%.2f km/h)  |  Distance: %.2f m",
                    gps_abs_vel_, car_speed_kmh_, total_distance_);
    }

    // ==========================================
    // ROS 2 Communication Handles
    // ==========================================
    rclcpp::Publisher<Float32>::SharedPtr speed_pub_;
    rclcpp::Publisher<Float32>::SharedPtr speed_pub_mps_;
    rclcpp::Subscription<NavSatFix>::SharedPtr gps_sub_;
    rclcpp::Subscription<TwistStamped>::SharedPtr gps_vel_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

    // ==========================================
    // Internal State Variables
    // ==========================================
    std::optional<double> prev_lat_;
    std::optional<double> prev_lon_;
    float total_distance_;
    float gps_abs_vel_;
    float car_speed_kmh_;
};

// ==========================================
// Main Execution Entry Point
// ==========================================
int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GPSDistanceVelocityNode>();

    try {
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_INFO(node->get_logger(), "❌ Node stopped by user.");
    }

    rclcpp::shutdown();
    return 0;
}

File: python_files/speed2.py
#!/usr/bin/env python3
"""
gps_distance_velocity_node.py

ROS 2 Node to compute:
- Ground (horizontal) speed from GPS velocity data
- Distance traveled between consecutive GPS coordinate updates
- Convert ground speed (m/s) to car speed (km/h)
- Publish current car speed on topic `/speed`

Author: Inzamam ul Haq
Date: October 2025
"""

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import NavSatFix
from geometry_msgs.msg import TwistStamped
from std_msgs.msg import Float32
import numpy as np


# -------------------------------------------------
# Utility: Speed conversion chart and function
# -------------------------------------------------
speed_chart: dict[str, float] = {
    "km/h": 1.0,
    "m/s": 3.6,
    "mph": 1.609344,
    "knot": 1.852,
}

speed_chart_inverse: dict[str, float] = {
    "km/h": 1.0,
    "m/s": 0.277777778,
    "mph": 0.621371192,
    "knot": 0.539956803,
}


def convert_speed(speed: float, unit_from: str, unit_to: str) -> float:
    """
    Convert speed from one unit to another using the charts above.
    Example:
      convert_speed(10, "m/s", "km/h") → 36.0
    """
    if unit_to not in speed_chart or unit_from not in speed_chart_inverse:
        msg = (
            f"Incorrect 'from_type' or 'to_type' value: {unit_from!r}, {unit_to!r}\n"
            f"Valid values are: {', '.join(speed_chart_inverse)}"
        )
        raise ValueError(msg)
    return round(speed * speed_chart[unit_from] * speed_chart_inverse[unit_to], 3)


# -------------------------------------------------
# Main Node Class
# -------------------------------------------------
class GPSDistanceVelocityNode(Node):
    def __init__(self):
        super().__init__('gps_distance_velocity_node')

        # ---- Variables ----
        self.prev_lat = None
        self.prev_lon = None
        self.total_distance = 0.0  # meters
        self.gps_abs_vel = 0.0     # m/s
        self.car_speed_kmh = 0.0   # km/h

        # ---- Publishers ----
        self.speed_pub = self.create_publisher(Float32, '/speed', 10)
        self.speed_pub_mps = self.create_publisher(Float32, '/speed_mps', 10)

        # ---- Subscribers ----
        self.create_subscription(
            NavSatFix,
            '/fix',
            self.gps_callback,
            10
        )

        self.create_subscription(
            TwistStamped,
            '/vel',
            self.gps_vel_callback,
            10
        )

        self.get_logger().info("✅ GPS Distance & Velocity Node Started")

        # ---- Timer for periodic logging ----
        self.timer = self.create_timer(2.0, self.timer_callback)

    # -------------------------------------------------
    # Callback: GPS Position
    # -------------------------------------------------
    def gps_callback(self, msg: NavSatFix):
        """Called whenever a new GPS fix message is received."""
        lat = msg.latitude
        lon = msg.longitude

        if self.prev_lat is not None and self.prev_lon is not None:
            dist = self.get_distance(self.prev_lat, self.prev_lon, lat, lon)
            self.total_distance += dist
            self.get_logger().info(
                f"📍 Distance from last point: {dist:.3f} m  |  Total: {self.total_distance:.3f} m"
            )

        # Update previous coordinates
        self.prev_lat = lat
        self.prev_lon = lon

    # -------------------------------------------------
    # Callback: GPS Velocity
    # -------------------------------------------------
    def gps_vel_callback(self, msg: TwistStamped):
        """Compute horizontal ground speed and convert to km/h."""
        vel_x = msg.twist.linear.x
        vel_y = msg.twist.linear.y

        # Ground speed (horizontal magnitude)
        self.gps_abs_vel = np.sqrt(vel_x ** 2 + vel_y ** 2)

        # Convert m/s → km/h
        self.car_speed_kmh = convert_speed(self.gps_abs_vel, "m/s", "km/h")

        # Publish the car speed in km/h
        speed_msg_mps = Float32()
        speed_msg_mps.data = self.gps_abs_vel
        self.speed_pub_mps.publish(speed_msg_mps)
        speed_msg = Float32()
        speed_msg.data = self.car_speed_kmh
        self.speed_pub.publish(speed_msg)

        self.get_logger().info(
            f"🚗 Ground Speed: {self.gps_abs_vel:.3f} m/s  |  Car Speed: {self.car_speed_kmh:.3f} km/h"
        )

    # -------------------------------------------------
    # Compute Distance between two GPS points
    # -------------------------------------------------
    def get_distance(self, lat1, lon1, lat2, lon2):
        """Returns distance in meters using the Haversine formula."""
        R = 6371.0 * 1000.0  # Earth radius in m

        lat_start = np.radians(lat1)
        lon_start = np.radians(lon1)
        lat_end = np.radians(lat2)
        lon_end = np.radians(lon2)

        dLat = lat_end - lat_start
        dLon = lon_end - lon_start

        a = (
            np.sin(dLat / 2.0) ** 2
            + np.cos(lat_start) * np.cos(lat_end) * np.sin(dLon / 2.0) ** 2
        )
        c = 2.0 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))
        d = c * R
        return d

    # -------------------------------------------------
    # Timer: Periodic summary
    # -------------------------------------------------
    def timer_callback(self):
        """Print current speed and total distance periodically."""
        self.get_logger().info(
            f"⏱️ Speed: {self.gps_abs_vel:.2f} m/s  ({self.car_speed_kmh:.2f} km/h)  |  Distance: {self.total_distance:.2f} m"
        )


# -------------------------------------------------
# Main Entry Point
# -------------------------------------------------
def main(args=None):
    rclpy.init(args=args)
    node = GPSDistanceVelocityNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('❌ Node stopped by user.')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

